torch/nn/parallel/distributed.py

```diff
@@ -563,6 +563,12 @@ def forward(self, *inputs, **kwargs):
         if self.require_forward_param_sync:
             self._sync_params()
 
+        # It may allocate new buckets before deallocating old buckets
+        # inside rebuildBuckets.
+        # To save peak memory usage, call rebuildBuckets before the peak
+        # memory usage increases during computing outputs.
+        self.reducer.rebuildBuckets()
```

Comment:
Since the launching of this function is moved to Python, I wonder if the conditioning should also be moved to Python (I mean `has_rebuilt_bucket_`)? Otherwise, this looks like a per-iteration overhead.

Reply:
I feel has_rebuilt_bucket_ should be kept in reducer, I can change the rebuildBucket name to prepare_forward() for better clarity